draco/analysis/beamform.py

```diff
@@ -0,0 +1,515 @@
+import numpy as np
+from caput import mpiarray, config, mpiutil, pipeline
+from ..core import task, containers, io
+from cora.util import units
+from ch_util import tools, ephemeris
+from drift.telescope import cylbeam
+from ..util._fast_tools import beamform
+from ..util.tools import polarization_map, baseline_vector
+from mpi4py import MPI
+
+# Constants
+NU21 = units.nu21
+C = units.c
+
+
+class BeamFormBase(task.SingleTask):
+    """ Base class for beam forming tasks.
+        Defines a few useful methods. Not to be used directly
+        but as parent class for BeamForm and BeamFormCat.
+
+    Attributes
+    ----------
+    collapse_ha : bool
+        Wether or not to sum over hour-angle/time to complete
+        the beamforming. Default is True, which sums over.
+    polarization : string
+        One of:
+        'I' : Stokes I only.
+        'full' : 'XX', 'YY', 'XY' and 'YX' in this order.
+        'copol' : 'XX' and 'YY' only.
+        'stokes' : 'I', 'Q', 'U' and 'V' in this order. Not implemented.
+    timetrack : float
+        How long (in seconds) to track sources at each side of transit.
+        Total transit time will be ~ 2 * timetrack.
+    """
+
+    collapse_ha = config.Property(proptype=bool, default=True)
+    polarization = config.Property(proptype=str, default='full')
+    timetrack = config.Property(proptype=float, default=900.)
+
+    def setup(self, manager):
+        """ Generic setup method. To be complemented by specific
+            setup methods in daughter tasks.
+
+        Parameters
+        ----------
+        manager : either `ProductManager`, `BeamTransfer` or `TransitTelescope`
+            Contains a TransitTelescope object describing the telescope.
+        """
+        # Get the TransitTelescope object
+        self.telescope = io.get_telescope(manager)
+        # Polarizations.
+        if self.polarization == 'I':
+            self.process_pol = ['XX', 'YY']
+            self.return_pol = ['I']
+        elif self.polarization == 'full':
+            self.process_pol = ['XX', 'YY', 'XY', 'YX']
+            self.return_pol = self.process_pol
+        elif self.polarization == 'copol':
+            self.process_pol = ['XX', 'YY']
+            self.return_pol = self.process_pol
+        elif self.polarization == 'stokes':
+            self.process_pol = ['XX', 'YY', 'XY', 'YX']
+            self.return_pol = ['I', 'Q', 'U', 'V']
+            msg = 'Stokes parameters are not implemented'
+            raise RuntimeError(msg)
+        else:
+            msg = 'Invalid polarization parameter: {0}'
+            msg = msg.format(self.polarization)
+            raise ValueError(msg)
+        # Number of polarizations to process
+        self.npol = len(self.process_pol)
+        self.latitude = np.deg2rad(ephemeris.CHIMELATITUDE)
```

Comment:
This is in `draco` it can't depend on CHIME. You can get the latitude from the `telescope` instance.